File: Duck.java

/**
 * Classe abstrata Duck - escreva a descrição da classe aqui
 * 
 * @author (seu nome aqui)
 * @version (versão ou data)
 */
public abstract class Duck
{
    protected QuackBehavior quackBehavior;
    
    public abstract void display();
    
    public void performQuack(){
        quackBehavior.quack();
    }
    
    public void setQuackBehavior(QuackBehavior q){
        quackBehavior = q;
    }
}

File: go_goodness/duck.go
package main

type Duck struct {
	quackBehavior QuackBehavior
}

func (d *Duck) display() {}

func (d *Duck) performQuack() {
	d.quackBehavior.quack()
}

func (d *Duck) setQuackBehavior(q QuackBehavior) {
	d.quackBehavior = q
}
